File: bogo2bogo/opencv/ImageViewer/ImageViewer.cpp
#include "ImageViewer.h"
#include "ui_ImageViewer.h"
#include <QtCore>
#include <QtGui>
#include <QtWidgets>

ImageSpinner::ImageSpinner(const QString &imagePath) : m_currIndex(-1)
{
   QFileInfo current(imagePath);
   m_dir = current.absoluteDir();
   QStringList imageFilters;
   imageFilters << "*.tiff"
                << "*.tif"
                << "*.jpg"
                << "*.jpeg"
                << "*.gif"
                << "*.bmp"
                << "*.png";
   m_fileNames = m_dir.entryList(imageFilters, QDir::Files, QDir::Name);
   m_currIndex = m_fileNames.indexOf(QRegExp(QRegExp::escape(current.fileName())));
}

QString ImageSpinner::prevImage()
{
   m_currIndex = m_currIndex - 1;
   if (m_currIndex < 0)
      m_currIndex = 0; // keep at 0th index
   QString prevImagePath = m_dir.absolutePath() + QDir::separator() + m_fileNames.at(m_currIndex);
   qDebug() << "Will display (prev): " << prevImagePath;
   //return dir.absoluteFilePath(m_fileNames.at(m_currIndex));
   return prevImagePath;
}

QString ImageSpinner::nextImage()
{
   m_currIndex = m_currIndex + 1;
   if (m_currIndex > m_fileNames.count() - 1)
      m_currIndex = m_fileNames.count() - 1; // keep at last index

   QString nextImagePath = m_dir.absolutePath() + QDir::separator() + m_fileNames.at(m_currIndex);
   qDebug() << "Will display (next): " << nextImagePath;
   //return dir.absoluteFilePath(m_fileNames.at(m_currIndex));
   return nextImagePath;
}

bool ImageSpinner::atFirst() const
{
   return m_currIndex == 0;
}

bool ImageSpinner::atLast() const
{
   return m_currIndex == m_fileNames.count() - 1;
}

ImageViewer::ImageViewer(QWidget *parent) : QMainWindow(parent), image(nullptr), imageSpinner(nullptr)
{
   setWindowTitle("ImageViewer");
   scaleFactor = 1.0;
   imageLoaded = false;
   imageLabel = new QLabel("");
   scrollArea = new QScrollArea();

   imageLabel->setBackgroundRole(QPalette::Base);
   imageLabel->setSizePolicy(QSizePolicy::Ignored, QSizePolicy::Ignored);
   imageLabel->setScaledContents(true);

   scrollArea->setBackgroundRole(QPalette::Dark);
   scrollArea->setWidget(imageLabel);
   scrollArea->setVisible(false);
   setCentralWidget(scrollArea);

   createActions();
   createMenus();
   createToolbars();
   statusBar()->showMessage("");
   statusBar()->setStyleSheet(
      "QStatusBar{padding-left:8px;background:rgb(59,59,59);color:rgb(127,127,127);}");

   // set initial size to 3/5 of screen
   resize(QGuiApplication::primaryScreen()->availableSize() * 4 / 5);
   setWindowIcon(QIcon(":/ImageViewer-icon.png")); // load an image
}

ImageViewer::~ImageViewer() {}

void ImageViewer::createActions()
{
   openAction = new QAction("&Open...", this);
   openAction->setShortcut(QKeySequence::Open);
   openAction->setIcon(QIcon(":/open.png"));
   openAction->setStatusTip("Open a new image file to view");
   QObject::connect(openAction, SIGNAL(triggered()), this, SLOT(open()));

   printAction = new QAction("&Print...", this);
   printAction->setShortcut(QKeySequence::Print);
   printAction->setIcon(QIcon(":/print.png"));
   printAction->setStatusTip("Print the current image");
   QObject::connect(printAction, SIGNAL(triggered()), this, SLOT(print()));
   printAction->setEnabled(false);

   exitAction = new QAction("E&xit", this);
   exitAction->setShortcut(QKeySequence("Ctrl+Q"));
   exitAction->setStatusTip("Quit the application");
   QObject::connect(exitAction, SIGNAL(triggered()), QApplication::instance(), SLOT(quit()));

   zoomInAction = new QAction("Zoom &in (25%)", this);
   zoomInAction->setShortcut(QKeySequence("Ctrl++"));
   zoomInAction->setIcon(QIcon(":/zoom_in.png"));
   zoomInAction->setStatusTip("Zoom into the image by 25%");
   QObject::connect(zoomInAction, SIGNAL(triggered()), this, SLOT(zoomIn()));
   zoomInAction->setEnabled(false);

   zoomOutAction = new QAction("Zoom &out (25%)", this);
   zoomOutAction->setShortcut(QKeySequence("Ctrl+-"));
   zoomOutAction->setIcon(QIcon(":/zoom_out.png"));
   zoomOutAction->setStatusTip("Zoom out of the image by 25%");
   QObject::connect(zoomOutAction, SIGNAL(triggered()), this, SLOT(zoomOut()));
   zoomOutAction->setEnabled(false);

   zoomNormalAction = new QAction("&Normal size", this);
   zoomNormalAction->setShortcut(QKeySequence("Ctrl+0"));
   zoomNormalAction->setStatusTip("Zoom to normal size (reset zoom)");
   QObject::connect(zoomNormalAction, SIGNAL(triggered()), this, SLOT(normalSize()));
   zoomNormalAction->setEnabled(false);

   fitToWindowAction = new QAction("Fit to &window", this);
   fitToWindowAction->setShortcut(QKeySequence("Ctrl+1"));
   fitToWindowAction->setIcon(QIcon(":/fit_to_size.png"));
   fitToWindowAction->setStatusTip("Fit the image to window");
   QObject::connect(fitToWindowAction, SIGNAL(triggered()), this, SLOT(fitToWindow()));
   fitToWindowAction->setEnabled(false);
   fitToWindowAction->setCheckable(true);

   prevImageAction = new QAction("&Previous image", this);
   prevImageAction->setShortcut(QKeySequence::MoveToPreviousChar); // left arrow (usually)
   prevImageAction->setIcon(QIcon(":/previous.png"));
   prevImageAction->setStatusTip("View previous image in folder");
   QObject::connect(prevImageAction, SIGNAL(triggered()), this, SLOT(prevImage()));
   prevImageAction->setEnabled(false);

   nextImageAction = new QAction("&Next image", this);
   nextImageAction->setShortcut(QKeySequence::MoveToNextChar); // right arrow (usually)
   nextImageAction->setIcon(QIcon(":/next.png"));
   nextImageAction->setStatusTip("View next image in folder");
   QObject::connect(nextImageAction, SIGNAL(triggered()), this, SLOT(nextImage()));
   nextImageAction->setEnabled(false);

   aboutAction = new QAction("&About...", this);
   aboutAction->setStatusTip("Display the about information");
   QObject::connect(aboutAction, SIGNAL(triggered()), this, SLOT(about()));

   aboutQtAction = new QAction("About &Qt...", this);
   aboutQtAction->setStatusTip("Display information about the Qt library");
   QObject::connect(aboutQtAction, SIGNAL(triggered()), QApplication::instance(), SLOT(aboutQt()));
}

void ImageViewer::createMenus()
{
   QMenu *fileMenu = menuBar()->addMenu("&File");
   fileMenu->addAction(openAction);
   fileMenu->addAction(printAction);
   fileMenu->addSeparator();
   fileMenu->addAction(exitAction);

   QMenu *viewMenu = menuBar()->addMenu("&View");
   viewMenu->addAction(zoomInAction);
   viewMenu->addAction(zoomOutAction);
   viewMenu->addAction(zoomNormalAction);
   viewMenu->addSeparator();
   viewMenu->addAction(fitToWindowAction);
   viewMenu->addSeparator();
   viewMenu->addAction(prevImageAction);
   viewMenu->addAction(nextImageAction);

   QMenu *helpMenu = menuBar()->addMenu("Help");
   helpMenu->addAction(aboutAction);
   helpMenu->addAction(aboutQtAction);
}

void ImageViewer::createToolbars()
{
   QToolBar *toolBar = addToolBar("&Main");
   QPalette palette = toolBar->palette();
   palette.setColor(QPalette::Window, QColor(59, 59, 59));
   toolBar->setPalette(palette);
   toolBar->addAction(openAction);
   toolBar->addAction(printAction);

   QToolBar *toolBar2 = addToolBar("&View");
   palette = toolBar2->palette();
   palette.setColor(QPalette::Window, QColor(59, 59, 59));
   toolBar2->addAction(zoomInAction);
   toolBar2->addAction(zoomOutAction);
   toolBar2->addAction(fitToWindowAction);
   toolBar2->addAction(prevImageAction);
   toolBar2->addAction(nextImageAction);
}

void ImageViewer::updateActions()
{
   fitToWindowAction->setEnabled(imageLoaded);
   zoomInAction->setEnabled(!fitToWindowAction->isChecked());
   zoomOutAction->setEnabled(!fitToWindowAction->isChecked());
   zoomNormalAction->setEnabled(!fitToWindowAction->isChecked());
   prevImageAction->setEnabled(imageLoaded);
   nextImageAction->setEnabled(imageLoaded);
}

void ImageViewer::scaleImage(double factor)
{
   scaleFactor *= factor;
   imageLabel->resize(scaleFactor * imageLabel->pixmap(Qt::ReturnByValue).size());

   adjustScrollBar(scrollArea->horizontalScrollBar(), factor);
   adjustScrollBar(scrollArea->verticalScrollBar(), factor);

   zoomInAction->setEnabled(scaleFactor < 5.0);
   zoomOutAction->setEnabled(scaleFactor > 0.10);
}

void ImageViewer::adjustScrollBar(QScrollBar *scrollBar, double factor)
{
   scrollBar->setValue(int(factor * scrollBar->value() + ((factor - 1) * scrollBar->pageStep() / 2)));
}

bool ImageViewer::loadImage(const QString &imagePath)
{
   QImageReader reader(imagePath);
   reader.setAutoTransform(true);
   const QImage newImage = reader.read();
   if (newImage.isNull()) {
      imageLoaded = false;
      QMessageBox::information(this, "ImageViewer", QString("FATAL: could not load image %1").arg(imagePath));
      return false;
   }

   scaleFactor = 1.0;
   imageLoaded = true;
   scrollArea->setVisible(true);
   imageLabel->setPixmap(QPixmap::fromImage(newImage));
   imageLabel->adjustSize();

   QString title;
   QTextStream ostr(&title);
   ostr << "Qt " << QT_VERSION_STR << " ImageViewer: " << imagePath;
   setWindowTitle(title);

   // add ImageSpinner
   delete imageSpinner;
   imageSpinner = new ImageSpinner(imagePath);

   return true;
}

void ImageViewer::initializeFileDialog(QFileDialog &dialog, QFileDialog::AcceptMode acceptMode)
{
   const QStringList picsLocation = QStandardPaths::standardLocations(QStandardPaths::PicturesLocation);
   dialog.setDirectory(picsLocation.isEmpty() ? QDir::currentPath() : picsLocation.last());

   QStringList mimeTypeFilters;
   const QByteArrayList supportedMimeTypes = (acceptMode == QFileDialog::AcceptOpen)
      ? QImageReader::supportedMimeTypes()
      : QImageWriter::supportedMimeTypes();
   for (const QByteArray &mimeTypeName : supportedMimeTypes)
      mimeTypeFilters.append(mimeTypeName);
   mimeTypeFilters.sort();
   dialog.setMimeTypeFilters(mimeTypeFilters);
   dialog.selectMimeTypeFilter("image/jpeg");
   dialog.setAcceptMode(acceptMode);
   if (acceptMode == QFileDialog::AcceptSave)
      dialog.setDefaultSuffix("jpg");
}

void ImageViewer::open()
{
   QFileDialog dialog(this, "Open Image");
   initializeFileDialog(dialog, QFileDialog::AcceptOpen);

   if (dialog.exec() == QFileDialog::Accepted && loadImage(dialog.selectedFiles().constFirst()))
      updateActions();
}

void ImageViewer::print()
{
   QMessageBox::information(this, "ImageViewer", "TODO: Implementation for Print Image");
}

void ImageViewer::zoomIn()
{
   scaleImage(1.25); // zoom in by 25%
}

void ImageViewer::zoomOut()
{
   scaleImage(0.75); // zoom out by 25%
}

void ImageViewer::normalSize()
{
   imageLabel->adjustSize();
   scaleFactor = 1.0;
}

void ImageViewer::fitToWindow()
{
   bool fitToWindow = fitToWindowAction->isChecked();
   scrollArea->setWidgetResizable(fitToWindow);
   if (!fitToWindow)
      normalSize();
   updateActions();
}

void ImageViewer::prevImage()
{
   QString imagePath = imageSpinner->prevImage();
   if (loadImage(imagePath))
      updateActions();
   if (imageSpinner->atFirst())
      QMessageBox::information(this, "ImageViewer", "Displaying first image in folder!");
}

void ImageViewer::nextImage()
{
   QString imagePath = imageSpinner->nextImage();
   if (loadImage(imagePath))
      updateActions();
   if (imageSpinner->atLast())
      QMessageBox::information(this, "ImageViewer", "Displaying last image in folder!");
}

void ImageViewer::about()
{
   QString str = QString("<p><b>Image Viewer</b> application to view images on desktop.</p>"
                         "<p>Developed with Qt %1 by Manish Bhobe</p>"
                         "<p>Free to use, but use at your own risk!!")
                    .arg(QT_VERSION_STR);
   QMessageBox::about(this, tr("About Image Viewer"), str);
}

File: bogo2bogo/opencv/ImageViewer/ImageViewer.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# ImageViewer.py: Image viewer application with PyQt
import sys
import os
import pathlib
from PyQt5.QtCore import *
from PyQt5.QtGui import *
from PyQt5.QtWidgets import *
from PyQt5 import uic
# using qdarkstyle (@see: https://github.com/ColinDuquesnoy/QDarkStyleSheet)
import qdarkstyle
# to detect dark themes (@see: https://pypi.org/project/darkdetect/)
import darkdetect
from ImageSpinner import ImageSpinner

# thisPath = pathlib.Path(__file__)
# print(thisPath.parents[0], thisPath.parents[1], os.path.join(thisPath.parents[1], 'common_files'))
sys.path.append(os.path.join(pathlib.Path(__file__).parents[2], 'common_files'))
from mypyqt5_utils import ThemeSetter


class ImageViewer(QMainWindow):
    def __init__(self, parent: QWidget = None):
        super().__init__(parent)
        self.imageLoaded = False
        self.imageLabel = QLabel("")
        self.scrollArea = QScrollArea()
        self.image = None
        self.scaleFactor = 1.0
        self.firstDialog = True
        self.imageSpinner = None
        self.setupUi()

    def setupUi(self):
        self.imageLabel.setBackgroundRole(QPalette.Base)
        self.imageLabel.setSizePolicy(QSizePolicy.Ignored, QSizePolicy.Ignored)
        self.imageLabel.setScaledContents(True)

        self.scrollArea.setBackgroundRole(QPalette.Dark)
        self.scrollArea.setWidget(self.imageLabel)
        self.scrollArea.setVisible(False)
        self.setCentralWidget(self.scrollArea)

        self.createActions()
        self.createMenu()
        self.createToolbar()
        # and status bar
        if darkdetect.isDark():
            self.statusBar().setStyleSheet(
                "QStatusBar{padding-left:8px;background:rgb(66,66,66);color:rgb(255,255,255);}")
        else:
            self.statusBar().setStyleSheet(
                "QStatusBar{padding-left:8px;background:rgb(240,240,240);color:rgb(54,54,54);}")
        self.statusBar().showMessage(
            f"Image Viewer: developed with PyQt {PYQT_VERSION_STR} by Manish Bhobe")

        self.resize(QGuiApplication.primaryScreen().availableSize() * (4 / 5))
        windowIconPath = os.path.join(ImageViewer.getImagesPath(), "ImageViewer-icon.png")
        self.setWindowIcon(QIcon(windowIconPath))

    def getImagesPath():
        p = pathlib.Path(__file__)
        images_path = os.path.join(os.path.split(str(p))[0], "images")
        return images_path

    def iconFromName(self, imageName, useDark=True):
        iconPath = os.path.join(ImageViewer.getImagesPath(), 'dark' if useDark else 'light')
        iconName = f"{imageName}24.png"
        iconPath = os.path.join(iconPath, iconName)
        assert os.path.exists(iconPath), f"{iconPath} - icon does not exist!"
        return QIcon(iconPath)

    def createActions(self):
        usingDarkTheme = darkdetect.isDark()
        # open image
        self.openAction = QAction("&Open...", self)
        self.openAction.setShortcut(QKeySequence.New)
        self.openAction.setIcon(self.iconFromName("open", usingDarkTheme))
        self.openAction.setStatusTip("Open a new image file to view")
        self.openAction.triggered.connect(self.open)

        # print image
        self.printAction = QAction("&Print...", self)
        self.printAction.setShortcut(QKeySequence.Print)
        self.printAction.setIcon(self.iconFromName("print", usingDarkTheme))
        self.printAction.setStatusTip("Print the current image")
        self.printAction.triggered.connect(self.print)
        self.printAction.setEnabled(False)

        # exit
        self.exitAction = QAction("E&xit", self)
        self.exitAction.setShortcut(QKeySequence("Ctrl+Q"))
        self.exitAction.setStatusTip("Exit the application")
        self.exitAction.triggered.connect(QApplication.instance().quit)

        # View category...
        self.zoomInAction = QAction("Zoom &in (25%)", self)
        self.zoomInAction.setShortcut(QKeySequence("Ctrl++"))
        self.zoomInAction.setIcon(self.iconFromName("zoomin", usingDarkTheme))
        self.zoomInAction.setStatusTip("Zoom into the image by 25%")
        self.zoomInAction.triggered.connect(self.zoomIn)
        self.zoomInAction.setEnabled(False)

        self.zoomOutAction = QAction("Zoom &out (25%)", self)
        self.zoomOutAction.setShortcut(QKeySequence("Ctrl++"))
        self.zoomOutAction.setIcon(self.iconFromName("zoomout", usingDarkTheme))
        self.zoomOutAction.setStatusTip("Zoom out of the image by 25%")
        self.zoomOutAction.triggered.connect(self.zoomOut)
        self.zoomOutAction.setEnabled(False)

        self.zoomNormalAction = QAction("&Normal size", self)
        self.zoomNormalAction.setShortcut(QKeySequence("Ctrl+0"))
        # self.zoomNormalAction.setIcon(QIcon("./images/zoom_out.png"))
        self.zoomNormalAction.setStatusTip("Zoom to normal size")
        self.zoomNormalAction.triggered.connect(self.zoomNormal)
        self.zoomNormalAction.setEnabled(False)

        self.fitToWindowAction = QAction("Fit to &window", self)
        self.fitToWindowAction.setShortcut(QKeySequence("Ctrl+1"))
        self.fitToWindowAction.setIcon(self.iconFromName("fit2window", usingDarkTheme))
        self.fitToWindowAction.setStatusTip("Fit image to size of window")
        self.fitToWindowAction.triggered.connect(self.fitToWindow)
        self.fitToWindowAction.setEnabled(False)
        self.fitToWindowAction.setCheckable(True)

        self.prevImageAction = QAction("&Previous Image", self)
        self.prevImageAction.setShortcut(QKeySequence.MoveToPreviousChar)
        self.prevImageAction.setIcon(self.iconFromName("prev", usingDarkTheme))
        self.prevImageAction.setStatusTip("View previous image in folder")
        self.prevImageAction.triggered.connect(self.prevImage)
        self.prevImageAction.setEnabled(False)

        self.nextImageAction = QAction("&Next Image", self)
        self.nextImageAction.setShortcut(QKeySequence.MoveToNextChar)
        self.nextImageAction.setIcon(self.iconFromName("next", usingDarkTheme))
        self.nextImageAction.setStatusTip("View next image in folder")
        self.nextImageAction.triggered.connect(self.nextImage)
        self.nextImageAction.setEnabled(False)

        # Help category
        self.aboutAction = QAction("&About...", self)
        self.aboutAction.setStatusTip("Display about application information")
        self.aboutAction.triggered.connect(self.about)

        self.aboutQtAction = QAction("About &Qt...", self)
        self.aboutQtAction.setStatusTip(
            "Display information about Qt library being used")
        self.aboutQtAction.triggered.connect(QApplication.instance().aboutQt)

    def createMenu(self):
        # file menu
        fileMenu = self.menuBar().addMenu("&File")
        fileMenu.addAction(self.openAction)
        fileMenu.addAction(self.printAction)
        fileMenu.addSeparator()
        fileMenu.addAction(self.exitAction)

        # view menu
        viewMenu = self.menuBar().addMenu("View")
        viewMenu.addAction(self.zoomInAction)
        viewMenu.addAction(self.zoomOutAction)
        viewMenu.addAction(self.zoomNormalAction)
        viewMenu.addSeparator()
        viewMenu.addAction(self.fitToWindowAction)
        viewMenu.addSeparator()
        viewMenu.addAction(self.prevImageAction)
        viewMenu.addAction(self.nextImageAction)

        # help menu
        helpMenu = self.menuBar().addMenu("Help")
        helpMenu.addAction(self.aboutAction)
        helpMenu.addAction(self.aboutQtAction)

    def createToolbar(self):
        toolBar = self.addToolBar("Main")
        palette = toolBar.palette()
        if darkdetect.isDark():
            palette.setColor(QPalette.Window, QColor(66, 66, 66))
        else:
            palette.setColor(QPalette.Window, QColor(240, 240, 240))
        toolBar.setPalette(palette)
        toolBar.addAction(self.openAction)
        toolBar.addAction(self.printAction)
        toolBar.addSeparator()
        toolBar.addAction(self.zoomInAction)
        toolBar.addAction(self.zoomOutAction)
        toolBar.addAction(self.fitToWindowAction)
        toolBar.addAction(self.prevImageAction)
        toolBar.addAction(self.nextImageAction)

    def updateActions(self):
        self.fitToWindowAction.setEnabled(not (self.image is None))
        self.zoomInAction.setEnabled(not self.fitToWindowAction.isChecked())
        self.zoomOutAction.setEnabled(not self.fitToWindowAction.isChecked())
        self.zoomNormalAction.setEnabled(not self.fitToWindowAction.isChecked())
        self.prevImageAction.setEnabled(not (self.image is None))
        self.nextImageAction.setEnabled(not (self.image is None))

    def displayImageInfo(self):
        if self.image.isNull():
            return
        print(f"Image info -> width: {self.image.width()} - height: {self.image.height()} " +
              f"- bits/pixel: {self.image.depth()}")

    def scaleImage(self, factor):
        self.scaleFactor *= factor
        self.imageLabel.resize(self.scaleFactor * self.imageLabel.pixmap().size())
        self.adjustScrollbar(self.scrollArea.horizontalScrollBar(), factor)
        self.adjustScrollbar(self.scrollArea.verticalScrollBar(), factor)
        self.zoomInAction.setEnabled(self.scaleFactor < 5.0)
        self.zoomOutAction.setEnabled(self.scaleFactor > 0.10)
        print(f"Scalefactor = {self.scaleFactor}")

    def adjustScrollbar(self, scrollBar: QScrollBar, factor):
        scrollBar.setValue(int(factor * scrollBar.value() +
                               ((factor - 1) * scrollBar.pageStep() / 2)))

    def initOpenDialog(self, dialog: QFileDialog, acceptMode: QFileDialog.AcceptMode):
        picLocations = QStandardPaths.standardLocations(QStandardPaths.PicturesLocation)
        # print(f"Standard pic locations: {picLocations}")
        dialog.setDirectory(QDir.currentPath if len(picLocations) == 0 else picLocations[-1])
        supportedMimeTypes = (QImageReader.supportedMimeTypes() if acceptMode == QFileDialog.AcceptMode.AcceptOpen
                              else QImageWriter.supportedMimeTypes())
        mimeTypeFilters = [str(mimeTypeName, 'utf-8') for mimeTypeName in supportedMimeTypes]
        mimeTypeFilters = sorted(mimeTypeFilters)
        dialog.setMimeTypeFilters(mimeTypeFilters)
        dialog.selectMimeTypeFilter("image/jpeg")
        dialog.setAcceptMode(acceptMode)
        if (acceptMode == QFileDialog.AcceptMode.AcceptSave):
            dialog.setDefaultSuffix("jpg")

    def loadImage(self, imagePath):
        assert os.path.exists(imagePath), f"FATAL: could not load file {filePath}"
        reader = QImageReader(imagePath)
        reader.setAutoTransform(True)
        self.image = reader.read()
        if self.image is None:
            QMessageBox.error(self, "ImageViewer", f"Could not load {imagePath}")
            return False

        self.scaleFactor = 1.0
        self.scrollArea.setVisible(True)
        self.imageLabel.setPixmap(QPixmap.fromImage(self.image))
        self.imageLabel.adjustSize()
        self.setWindowTitle(f"PyQt {PYQT_VERSION_STR} Image Viewer: {imagePath}")
        self.fitToWindow()

        # image spinner
        if (self.imageSpinner is not None):
            del self.imageSpinner
            self.imageSpinner = None
        self.imageSpinner = ImageSpinner(imagePath)

        return True

    def open(self):
        openDialog = QFileDialog(self, "Open Image")
        self.initOpenDialog(openDialog, QFileDialog.AcceptMode.AcceptOpen)
        if openDialog.exec():
            filePath = list(openDialog.selectedFiles())[-1]
            if self.loadImage(filePath):
                self.updateActions()
                self.displayImageInfo()

    def print(self):
        QMessageBox.information(self, "ImageViewer",
                                "This is the 'print' action handler - yet to be implemented")

    def zoomIn(self):
        self.scaleImage(1.25)

    def zoomOut(self):
        self.scaleImage(0.75)

    def zoomNormal(self):
        self.imageLabel.adjustSize()
        self.scaleFactor = 1.0

    def fitToWindow(self):
        fitToWindow = self.fitToWindowAction.isChecked()
        self.scrollArea.setWidgetResizable(fitToWindow)
        if not fitToWindow:
            self.zoomNormal()
        self.updateActions()

    def prevImage(self):
        assert(self.imageSpinner is not None)
        imagePath = self.imageSpinner.prevImagePath()
        if (self.loadImage(imagePath)):
            self.updateActions()
        if (self.imageSpinner.atFirstPath()):
            QMessageBox.information(self, "ImageViewer", "Displaying first image in folder!")

    def nextImage(self):
        assert(self.imageSpinner is not None)
        imagePath = self.imageSpinner.nextImagePath()
        if (self.loadImage(imagePath)):
            self.updateActions()
        if (self.imageSpinner.atLastPath()):
            QMessageBox.information(self, "ImageViewer", "Displaying last image in folder!")

    def about(self):
        QMessageBox.about(self, "About Image Viewer",
                          f"<p><b>Image Viewer</b> application to view images on desktop.</p>"
                          f"<p>Developed with PyQt {PYQT_VERSION_STR} by Manish Bhobe</p>"
                          f"<p>Free to use, but use at your own risk!!")


def main():
    app = QApplication(sys.argv)
    # app.setFont(QApplication.font("QMenu"))
    app.setStyle("Fusion")
    # palSwitcher = utils.PaletteSwitcher(app)

    if darkdetect.isDark():
        ThemeSetter.setDarkTheme(app)
    else:
        ThemeSetter.setLightTheme(app)

    w = ImageViewer()
    w.setWindowTitle(f"PyQt {PYQT_VERSION_STR} Image Viewer")
    w.show()

    sys.exit(app.exec())


if __name__ == '__main__':
    main()
